src/mapgen.cpp

```diff
@@ -199,27 +199,86 @@ void Mapgen::updateHeightmap(v3s16 nmin, v3s16 nmax)
 	//printf("updateHeightmap: %dus\n", t.stop());
 }
 
+inline bool Mapgen::updateLiquidHelper(int i, const v3s16 &em)
+{
+	u32 i_nx = i;
+	vm->m_area.add_x(em, i_nx, -1);
+	if (vm->m_data[i_nx].getContent() != CONTENT_IGNORE) {
+		const ContentFeatures &c_nx = ndef->get(vm->m_data[i_nx]);
+		if (c_nx.floodable && !c_nx.isLiquid())
+			return true;
+	}
+	u32 i_px = i;
+	vm->m_area.add_x(em, i_px, +1);
+	if (vm->m_data[i_px].getContent() != CONTENT_IGNORE) {
+		const ContentFeatures &c_px = ndef->get(vm->m_data[i_px]);
+		if (c_px.floodable && !c_px.isLiquid())
+			return true;
+	}
+	u32 i_nz = i;
+	vm->m_area.add_z(em, i_nz, -1);
+	if (vm->m_data[i_nz].getContent() != CONTENT_IGNORE) {
+		const ContentFeatures &c_nz = ndef->get(vm->m_data[i_nz]);
+		if (c_nz.floodable && !c_nz.isLiquid())
+			return true;
+	}
+	u32 i_pz = i;
+	vm->m_area.add_z(em, i_pz, +1);
+	if (vm->m_data[i_pz].getContent() != CONTENT_IGNORE) {
+		const ContentFeatures &c_pz = ndef->get(vm->m_data[i_pz]);
+		if (c_pz.floodable && !c_pz.isLiquid())
+			return true;
+	}
+	return false;
+}
 
 void Mapgen::updateLiquid(UniqueQueue<v3s16> *trans_liquid, v3s16 nmin, v3s16 nmax)
 {
-	bool isliquid, wasliquid;
+	bool isignored, isliquid, wasignored, wasliquid, waschecked, waspushed;
 	v3s16 em  = vm->m_area.getExtent();
 
-	for (s16 z = nmin.Z; z <= nmax.Z; z++) {
-		for (s16 x = nmin.X; x <= nmax.X; x++) {
-			wasliquid = true;
-
-			u32 i = vm->m_area.index(x, nmax.Y, z);
-			for (s16 y = nmax.Y; y >= nmin.Y; y--) {
-				isliquid = ndef->get(vm->m_data[i]).isLiquid();
-
-				// there was a change between liquid and nonliquid, add to queue.
-				if (isliquid != wasliquid)
+	for (s16 z = nmin.Z + 1; z <= nmax.Z - 1; z++)
+	for (s16 x = nmin.X + 1; x <= nmax.X - 1; x++) {
+		wasignored = true;
+		wasliquid = false;
+		waschecked = false;
+		waspushed = false;
+
+		u32 i = vm->m_area.index(x, nmax.Y, z);
```

Comment:
For what it's worth, the convention is to use `vi` as the name of a variable that represents the index of a node inside a VoxelManip.